### PR TITLE
update NumberDataPoint to support DoubleVal and IntVal

### PR DESCRIPTION
**Pull request by codeboten**
**Description:** updating NumberDataPoint to support both Double and Int, this is a stepping stone to remove IntDataPoint altogether.

**Link to tracking Issue:** Part of #3534

### CODE REVIEW

File: model/pdata/generated_metrics.go [hunk 1]
```diff
@@ -1203,7 +1215,9 @@ func (ms NumberDataPoint) CopyTo(dest NumberDataPoint) {
 	ms.LabelsMap().CopyTo(dest.LabelsMap())
 	dest.SetStartTimestamp(ms.StartTimestamp())
 	dest.SetTimestamp(ms.Timestamp())
-	dest.SetValue(ms.Value())
+	dest.SetDoubleVal(ms.DoubleVal())
+	dest.SetIntVal(ms.IntVal())
```
Comment (bogdandrutu) [inline model/pdata/generated_metrics.go:1219]: This does not seem correct, since you always set to be int value.

File: model/pdata/metrics.go [hunk 2]
```diff
@@ -305,3 +305,37 @@ const (
...
+
+// NewNumberDataPointValueInt creates a new NumberDataPointValue with the given int64 value.
+func NewNumberDataPointValueInt(v int64) NumberDataPoint {
+	return NumberDataPoint{orig: &otlpmetrics.NumberDataPoint{Value: &otlpmetrics.NumberDataPoint_AsInt{AsInt: v}}}
+}
+
+// NewNumberDataPointValueDouble creates a new NumberDataPointValue with the given float64 value.
+func NewNumberDataPointValueDouble(v float64) NumberDataPoint {
+	return NumberDataPoint{orig: &otlpmetrics.NumberDataPoint{Value: &otlpmetrics.NumberDataPoint_AsDouble{AsDouble: v}}}
+}
```
Comment (bogdandrutu) [inline model/pdata/metrics.go:341]: No need for these 2 constructors yet correct?

File: cmd/pdatagen/internal/base_fields.go [hunk 3]
```diff
@@ -462,3 +474,77 @@ func (pf *primitiveAsDoubleField) generateCopyToValue(sb *strings.Builder) {
...
+	for _, field := range pf.fields {
+		sb.WriteString("\ttv.Set" + field.name + "(" + field.testVal + ")\n")
+	}
+
+}
+
+func (pf *numberField) generateCopyToValue(sb *strings.Builder) {
+	for _, field := range pf.fields {
+		sb.WriteString("\tdest.Set" + field.name + "(ms." + field.name + "())\n")
+	}
```
Comment (bogdandrutu) [inline cmd/pdatagen/internal/base_fields.go:547]: This should be switch on the "type" and call set only for that.

File: cmd/pdatagen/internal/base_fields.go [hunk 4]
```diff
@@ -462,3 +474,77 @@ func (pf *primitiveAsDoubleField) generateCopyToValue(sb *strings.Builder) {
...
+
+type field struct {
+	name            string
+	defaultVal      string
+	testVal         string
+	returnType      string
+	originFieldName string
+	originFullName  string
+	asName          string
+}
```
Comment (bogdandrutu) [inline cmd/pdatagen/internal/base_fields.go:486]: Name this struct: "oneOfPrimitiveValue"

File: cmd/pdatagen/internal/base_fields.go [hunk 5]
```diff
@@ -462,3 +474,77 @@ func (pf *primitiveAsDoubleField) generateCopyToValue(sb *strings.Builder) {
...
+				return field.originFullName
+			case "asName":
+				return field.asName
+			default:
+				panic(name)
+			}
+		}))
+		sb.WriteString("\n")
+	}
+}
```
Comment (bogdandrutu) [inline cmd/pdatagen/internal/base_fields.go:515]: I would split the current `generateAccessors` into 2:

```golang
func (opv *oneofPrimitiveValue) generateAccessors(ms baseStruct, sb *strings.Builder) {
	sb.WriteString(os.Expand(accessorsNumberTemplate, func(name string) string {
		switch name {
		case "structName":
		return ms.getName()
		case "fieldName":
			return field.name
		case "lowerFieldName":
			return strings.ToLower(field.name)
		case "returnType":
			return field.returnType
		case "originFieldName":
			return field.originFieldName
		case "originFullName":
			return field.originFullName
		case "asName":
			return field.asName
		default:
			panic(name)
		}
	}))
	sb.WriteString("\n")
}

func (nf *numberField) generateAccessors(ms baseStruct, sb *strings.Builder) {
	for _, field := range nf.fields {
		field.generateAccessors(ms, sb)
	}
}
```

File: cmd/pdatagen/internal/base_fields.go [hunk 6]
```diff
@@ -462,3 +474,77 @@ func (pf *primitiveAsDoubleField) generateCopyToValue(sb *strings.Builder) {
...
+type field struct {
+	name            string
+	defaultVal      string
+	testVal         string
+	returnType      string
+	originFieldName string
+	originFullName  string
+	asName          string
+}
+type numberField struct {
```
Comment (bogdandrutu) [inline cmd/pdatagen/internal/base_fields.go:487]: Rename: `oneofField`

File: model/pdata/metrics.go (hunk 2 above)
Comment (codeboten) [reply model/pdata/metrics.go:341]: Correct, I can remove until they're needed.

File: cmd/pdatagen/internal/base_fields.go (hunk 6 above)
Comment (codeboten) [reply cmd/pdatagen/internal/base_fields.go:487]: This would conflict w/ the existing `oneofField`, are you suggesting this would replace that?

Comment (bogdandrutu) [reply cmd/pdatagen/internal/base_fields.go:487]: Eventually yes, but let's not resolve the world now :)), so maybe keep the current name number... and then we can look into replacing that :)

File: cmd/pdatagen/internal/base_fields.go [hunk 7]
```diff
@@ -462,3 +474,87 @@ func (pf *primitiveAsDoubleField) generateCopyToValue(sb *strings.Builder) {
...
+}
+
+var _ baseField = (*numberField)(nil)
+
+// switch ms.Type() {
+// case NumberDataPointTypeDouble:
+// 	dest.SetDoubleVal(ms.DoubleVal())
+// case NumberDataPointTypeInt:
+// 	dest.SetIntVal(ms.IntVal())
+// }
```
Comment (bogdandrutu) [inline cmd/pdatagen/internal/base_fields.go:560]: What is this? 🗡️

File: cmd/pdatagen/internal/base_fields.go (hunk 5 above)
Comment (codeboten) [reply cmd/pdatagen/internal/base_fields.go]: Done, I did the same for `generateAccessorsTest`

File: cmd/pdatagen/internal/base_fields.go (hunk 4 above)
Comment (codeboten) [reply cmd/pdatagen/internal/base_fields.go]: Done

File: cmd/pdatagen/internal/base_fields.go (hunk 3 above)
Comment (codeboten) [reply cmd/pdatagen/internal/base_fields.go]: Yup, fixed.

File: cmd/pdatagen/internal/base_fields.go [hunk 8]
```diff
@@ -462,3 +474,89 @@ func (pf *primitiveAsDoubleField) generateCopyToValue(sb *strings.Builder) {
...
+			return opv.testVal
+		default:
+			panic(name)
+		}
+	}))
+	sb.WriteString("\n")
+}
+
+type numberField struct {
+	fields []oneOfPrimitiveValue
```
Comment (bogdandrutu) [inline cmd/pdatagen/internal/base_fields.go:531]: Use slice of pointers here.

File: cmd/pdatagen/internal/base_fields.go [hunk 9]
```diff
@@ -462,3 +474,89 @@ func (pf *primitiveAsDoubleField) generateCopyToValue(sb *strings.Builder) {
...
+		sb.WriteString("\ttv.Set" + field.name + "(" + field.testVal + ")\n")
+	}
+
+}
+
+func (nf *numberField) generateCopyToValue(sb *strings.Builder) {
+	sb.WriteString("\tswitch ms.Type() {\n")
+	for _, field := range nf.fields {
+		sb.WriteString("\tcase NumberDataPointType" + field.fieldType + ":\n")
+		sb.WriteString("\t dest.Set" + field.name + "(ms." + field.name + "())\n")
```
Comment (bogdandrutu) [inline cmd/pdatagen/internal/base_fields.go:557]: This line can be in a func "func (opv *oneOfPrimitiveValue) generateCopyToValue(sb *strings.Builder)" so we can make smarter things later :)

File: cmd/pdatagen/internal/base_fields.go [hunk 10]
```diff
@@ -462,3 +474,89 @@ func (pf *primitiveAsDoubleField) generateCopyToValue(sb *strings.Builder) {
...
+
+func (nf *numberField) generateAccessorsTest(ms baseStruct, sb *strings.Builder) {
+	for _, field := range nf.fields {
+		field.generateAccessorsTest(ms, sb)
+	}
+}
+
+func (nf *numberField) generateSetWithTestValue(sb *strings.Builder) {
+	for _, field := range nf.fields {
+		sb.WriteString("\ttv.Set" + field.name + "(" + field.testVal + ")\n")
```
Comment (bogdandrutu) [inline cmd/pdatagen/internal/base_fields.go:548]: This line can be in a func "func (opv *oneOfPrimitiveValue) generateSetWithTestValue(sb *strings.Builder)" so we can make smarter things later :)

File: cmd/pdatagen/internal/base_fields.go (hunk 9 above)
Comment (codeboten) [reply cmd/pdatagen/internal/base_fields.go]: Done!

File: cmd/pdatagen/internal/base_fields.go (hunk 10 above)
Comment (codeboten) [reply cmd/pdatagen/internal/base_fields.go]: Done!

File: cmd/pdatagen/internal/base_fields.go (hunk 8 above)
Comment (codeboten) [reply cmd/pdatagen/internal/base_fields.go]: Fixed!

File: model/pdata/generated_metrics_test.go [hunk 11]
```diff
@@ -2007,7 +2014,9 @@ func fillTestNumberDataPoint(tv NumberDataPoint) {
 	fillTestStringMap(tv.LabelsMap())
 	tv.SetStartTimestamp(Timestamp(1234567890))
 	tv.SetTimestamp(Timestamp(1234567890))
-	tv.SetValue(float64(17.13))
+	tv.SetDoubleVal(float64(17.13))
+	tv.SetIntVal(int64(17))
```
Comment (bogdandrutu) [inline model/pdata/generated_metrics_test.go:2018]: Should this be only one set?

File: model/pdata/metrics.go [hunk 12]
```diff
@@ -305,3 +305,39 @@ const (
 func (at AggregationTemporality) String() string {
 	return otlpmetrics.AggregationTemporality(at).String()
 }
+
+// NumberDataPointType specifies the type of NumberDataPoint.
+type NumberDataPointType int32
+
+const (
+	NumberDataPointTypeNull NumberDataPointType = iota
```
Comment (bogdandrutu) [inline model/pdata/metrics.go:313]: s/NumberDataPointTypeNull/NumberDataPointTypeNone?

**Review by bogdandrutu — COMMENTED** [review]
LGTM only 2 minor comments

File: model/pdata/metrics.go (hunk 12 above)
Comment (codeboten) [reply model/pdata/metrics.go:313]: Could go w/ None, I followed the naming convention used by AttributeValue `AttributeValueTypeNull`

Comment (bogdandrutu) [reply model/pdata/metrics.go:313]: Let's go with None for this, makes more sense. I will think about AttributeValue

File: model/pdata/generated_metrics_test.go (hunk 11 above)
Comment (codeboten) [reply model/pdata/generated_metrics_test.go]: SetDoubleVal for now, added a TODO to remember to test with both int and double values in the future.

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**